### ISSUE TITLE
hulu.com, viz.com

### ISSUE THREAD

**Issue by jspenguin2017** [jspenguin2017/uBlockProtector#398]
https://www.hulu.com/watch/132768?d=AnimePlanet
https://www.viz.com/watch/streaming/bleach-episode-366/4380/dub

No screenshot for `hulu.com` as it requires account first. 
![image](https://user-images.githubusercontent.com/7283682/28234729-3ab32eae-68c0-11e7-9721-b4b130ae6ca8.png)

https://gitlab.com/xuhaiyang1234/AAK-Cont/issues/63

**Comment by uBlock-user** [comment]
hulu.com requires Flash too!

**Comment by uBlock-user** [comment]
Anti adblock script for `viz.com` https://www.viz.com/assets/manifest-viz-streaming-5aba0337ce23a74275b63d59bef0e1eab16e924b48b26d1ba8d0ea428f1793dd.js

**Comment by uBlock-user** [comment]
List of setTimeout calls made by the script - https://gist.githubusercontent.com/uBlock-user/2c80fa1e6f597b325f5b94e1aa479469/raw/b8fdc5114e162a758920241ac6493510ab7ea1b3/gistfile1.txt

**Comment by uBlock-user** [comment]
That should kill the warning and the player hiding js

```
viz.com##script:inject(setTimeout-defuser.js, Please disable the ad blocker and try reloading this page)
viz.com##script:inject(setTimeout-defuser.js, Please try disabling ad blockers and reloading this page)
```

**Comment by uBlock-user** [comment]
I'm not sure if `viz.com` also requires Flash, need some feedback!

**Comment by ghajini** [comment]
someone in usa need to check that if it detects vpn....

**Comment by jspenguin2017** [comment]
Are you sure that works? They have a wrapper that hides the function.

**Comment by jspenguin2017** [comment]
Anyway, I fixed the message in another way, but video still doesn't play, even if I enable Flash.

**Comment by ghajini** [comment]
i see no video panel....only blank space

**Comment by jspenguin2017** [comment]
Yea, I am aware of that.

**Comment by uBlock-user** [comment]
Works as in those adblocking messages no longer appear as the detection is killed with the call, the video not playing is another issue.

**Comment by jspenguin2017** [comment]
It's not really another issue as it plays with uBO off.

**Comment by jspenguin2017** [comment]
Actually not quite, since I'm Geo Locked.

**Comment by uBlock-user** [comment]
another issue meaning not related to anti adblock, something else needs to be patched too or is blocked which means it's broken.

**Comment by uBlock-user** [comment]
Not playing the video even after disabling all my extensions and uBO.

**Comment by uBlock-user** [comment]
Found this in the console - **A Parser-blocking, cross site (i.e. different eTLD+1) script, https://player.hulu.com/js/hulu_global_html5.js?guid=0B1CEA04-DCCB-40cf-AD0E-5222EF66D519&partner=VizAnime&wmode=opaque&width=100%25&height=100%25, is invoked via document.write. The network request for this script MAY be blocked by the browser in this or a future page load due to poor network connectivity. If blocked in this page load, it will be confirmed in a subsequent console message.See https://www.chromestatus.com/feature/5718547946799104 for more details.**

That is probably breaking the Hulu embedded player.

**Comment by uBlock-user** [comment]
It appears the browser itself is blocking the player itself. Anyone have any other browser should try loading the URL there.

**Comment by jspenguin2017** [comment]
Synchronous cross origin request are not blocked unless you are on 2G network.

**Comment by jspenguin2017** [comment]
![image](https://user-images.githubusercontent.com/7283682/28240749-3ecd170a-6944-11e7-87ef-daa3c578ed5b.png)

**Comment by jspenguin2017** [comment]
![image](https://user-images.githubusercontent.com/7283682/28240755-552446f4-6944-11e7-86e1-a994afa2eeaa.png)

**Comment by jspenguin2017** [comment]
Synchronous XHR is actually never blocked on Wi-Fi, when I was testing Violentproxy, there was a misconfiguration that is causing POST requests to be dropped, and some webpages just lock up.

**Comment by uBlock-user** [comment]
I'm not on a 2G network obviously and it still happening on my end. I don't see that notice either.

**Comment by jspenguin2017** [comment]
That notice will always show, but... 
> If blocked in this page load, it will be confirmed in a subsequent console message.`

Do you see the subsequent message? If not, then it's not blocked.

**Comment by uBlock-user** [comment]
By notice I meant the Geo-Lock notice and I don't see any subsequent messages related to that.

**Comment by jspenguin2017** [comment]
Ah, I read it wrong... I think it only show when you have Flash...

**Comment by jspenguin2017** [comment]
I still can't find a way to debug it...

**Comment by ghajini** [comment]
really need a user/contributor who is using usa isp///

**Comment by jspenguin2017** [comment]
Most likely a contributor, an user is unlikely to be able to debug it.

**Comment by ghajini** [comment]
hey friends i can provide trick describe for free hulu account
https://www.youtube.com/watch?v=3QmYvIGMR80

& regarding geolock still don't find any way other than a contributor/user with us isp

**Comment by ghajini** [comment]
hurray!!!!!!
iam able to pass restriction (of using proxy tool).....
just try this dns trick   https://unlocator.com/setup/   depending on your os
and then restart your wifi & pc/laptop


https://streamable.com/j4wiv

**Comment by ghajini** [comment]
@jspenguin2017 
confidential issue is opened in aak cont
something useful 
https://gitlab.com/xuhaiyang1234/AAK-Cont/issues/182

**Comment by jspenguin2017** [comment]
Handled. 
Doesn't quite work.

**Comment by ghajini** [comment]
aak cont thread updated,plz see

**Comment by ghajini** [comment]
i provided enough info...i think you should start slowly

**Comment by jspenguin2017** [comment]
I'll look into it eventually, if you have a solution that'll speed it up.

**Comment by ghajini** [comment]
no i just provided you ...
solution for `detects vpn`
                    `account required`
                      `geo lock`

so that you can debug easily....

**Comment by jspenguin2017** [comment]
If it's so easy why don't you do it?

**Comment by ghajini** [comment]
i know its not easy but seeing at this thread and your comments i see you stuck due to vpn detect restrictions,account restrictions...so i just provided that.... i have very small knowledge of js ... i am not burdening on you?
thank you

**Comment by jspenguin2017** [comment]
I have the labels so that if someone comes to help, he can easily see what potential problems he will encounter.

**Comment by jspenguin2017** [comment]
Does the latest version works? Test links: 
https://www.hulu.com/watch/132768?d=AnimePlanet
https://www.viz.com/watch/streaming/bleach-episode-366/4380/dub

**Comment by uBlock-user** [comment]
I don't see the anti adblock notice anymore on both, but the video doesn't load on viz. Does it require flash too like hulu ?

**Comment by jspenguin2017** [comment]
Yea, require Flash and Geo Locked.

**Comment by ghajini** [comment]
on viz i see antiadblock....
eg on video startup for upto 10 sec ,antiadblock notice is shown

<details><summary>Screenshot:</summary>

![image](%https://user-images.githubusercontent.com/20338483/30248363-20f8e01a-9644-11e7-8223-2e7a61ca18ae.png%)
</details><br/>
 then video starts.....

at reaching white dots on video control bar,again antiadblock notice is shown in a set of 4 countdowns
<details><summary>Screenshot:</summary>

![image](%https://user-images.githubusercontent.com/20338483/30248376-58d6b3c2-9644-11e7-9e63-6310d3e4d261.png%)
</details><br/>

<details><summary>Screenshot:</summary>

![image](%https://user-images.githubusercontent.com/20338483/30248377-58d9f9c4-9644-11e7-8b35-bee995889475.png%)
</details><br/>

<details><summary>Screenshot:</summary>

![image](%https://user-images.githubusercontent.com/20338483/30248378-58e2ff88-9644-11e7-836e-e77394ddc98d.pngl%)
</details><br/>

for hulu account trick is not working,so will inform when i get

**Comment by jspenguin2017** [comment]
OK, I got the DNS to work. The ads requests are within the Flash object itself, I have no way to interfere with it. 
![image](https://user-images.githubusercontent.com/7283682/30249730-4b4a97c2-95ff-11e7-8cfc-5cb0d1db89bf.png)

**Comment by ghajini** [comment]
instead of preroll/midroll ad they show antiadblock and that too with timer(1-2min) for 4 sets of ads....how its 'acceptable fix' by the way?
sorry iam not arguing,but its your choice

**Comment by ghajini** [comment]
these adblock warning may be due to blocking of `moatads`, can you loopback it?( i locally whitelisted moatads and i see no warning but ads)

**Comment by jspenguin2017** [comment]
You got a link?

**Comment by jspenguin2017** [comment]
The requested file is a Flash object, I can't loopback that.

**Comment by jspenguin2017** [comment]
Anyway, if someone want to tinker around, here is a Node script that might be useful: 
<details>

```JavaScript
//Temporary DNS redirector
"use strict";

/**
 * Custom DNS servers.
 * @const {Array.<DNSServer>}
 */
const DNS_SERVERS = ["185.37.37.37", "185.37.39.39"]; //https://unlocator.com/

/**
 * Load modules.
 * @const {Module}
 */
const http = require("http");
const https = require("https");
const url = require("url");
const dns = require("dns");
const net = require("net");

/**
 * Regular expression to test if a string is filled with digits.
 * @const {RegExp}
 */
const reIsAllDigits = /^\d+$/;

//Set DNS server
dns.setServers(DNS_SERVERS);

//Start server
http.createServer().listen(12345).on("request", (req, res) => {
    //HTTP requests do not go though custom DNS for now, and insecure WebSocket does not work
    console.log("REQUEST", req.url);

    let option;
    try {
        option = url.parse(req.url);
    } catch (err) {
        res.writeHead(400, "Bad Request");
        res.end();
        return;
    }

    if (option.protocol !== "http:" && option.protocol !== "https:") {
        res.writeHead(400, "Bad Request");
        res.end();
        return;
    }

    delete req.headers["proxy-authorization"];
    if (req.headers["proxy-connection"]) {
        req.headers["connection"] = req.headers["proxy-connection"];
        delete req.headers["proxy-connection"];
    }

    option.method = req.method;
    option.headers = req.headers;

    const remote = (option.protocol === "http:" ? http : https).request(option, (rres) => {
        res.writeHead(rres.statusCode, rres.statusMessage, rres.headers);
        rres.pipe(res);
    });
    req.pipe(remote);

    remote.on("error", () => {
        res.destroy();
    });
    req.on("error", () => {
        remote.destroy();
    });
}).on("connect", (req, socket, head) => {
    //Everything works as expected in HTTPS mode
    console.log("CONNECT", req.url);

    let port = 443;
    let domain = "";
    {
        let parts = req.url.split(':');
        if (reIsAllDigits.test(parts[parts.length - 1])) {
            port = parseInt(parts.pop());
            if (port < 0 || port > 65535) {
                port = 443;
            }
        }
        domain = parts.join(':');
    }

    dns.resolve(domain, (err, address) => {
        let ip;
        if (err) {
            ip = domain;
        } else {
            ip = address[0];
            console.log(`\t${domain} -> ${ip}`);
        }

        const remote = net.connect({
            host: ip,
            port: port,
            servername: domain,
        }, () => {
            socket.write("HTTP/" + req.httpVersion + " 200 Connection Established\r\n");
            if (req.headers["proxy-connection"] === "keep-alive") {
                socket.write("Connection: keep-alive\r\n");
            }
            socket.write("\r\n");

            remote.write(head);

            socket.pipe(remote);
            remote.pipe(socket);
        });

        remote.on("error", () => {
            socket.destroy();
        });
        socket.on("error", () => {
            remote.destroy();
        });
    });
});

//Start up log
console.log("Server started on http://localhost:12345/");
```

</details>

Start it and use `http://localhost:12345/` as proxy.

**Comment by ghajini** [comment]
How to use this? Any manual pls?

**Comment by jspenguin2017** [comment]
It's for developers.

**Comment by ghajini** [comment]
Does v12 accomplishments help here? Like replace or mp4 modifier rule or anything?

**Comment by jspenguin2017** [comment]
From my previous testing, it seems that `viz.com` is no longer embedding `hulu.com`, so the issue is gone by itself. I don't have a way to test `hulu.com` itself, but I have a way to simulate `$mp4` modifier.

**Comment by ghajini** [comment]
You require hulu account(check if any cookie there on jaguartrials.net) or anything else(VPN,try vyprvpn with sanfrancisco server,chameleon technology)...

**Comment by jspenguin2017** [comment]
The Geo Lock is not a problem. I do need a test account.

**Comment by ghajini** [comment]
Does this work 
https://jaguartrials.net/cookies.php?id=238

**Comment by jspenguin2017** [comment]
Not sure, I might try it later.

**Comment by jspenguin2017** [comment]
Alright, replaced all the ads with a 1 second blank video. 
Should work now, might need some testing.

**Comment by uBlock-user** [comment]
Did they do it again ?

**Comment by jspenguin2017** [comment]
Do what?

**Comment by rogersachan** [comment]
@jspenguin2017 Appears that this is broken as of 9/9

**Comment by jspenguin2017** [comment]
RIP... It's horrifyingly hard to test these issues. I'll see when I can have another look at it. 
Also: https://github.com/NanoAdblocker/NanoFilters/issues/97
